### PR TITLE
Custom metrics

### PR DESCRIPTION
**Pull request by xRay2016**
The implementation of custom metrics.

### CODE REVIEW

File: contrib/pgexporter.conf [hunk 1]
```diff
@@ -0,0 +1,15 @@
+[pgexporter]
```
Comment (jesperpedersen) [inline contrib/pgexporter.conf:1]: Lets delete this file.

We need to have documentation in `doc/CONFIGURATION.md` about `metrics_path`

File: contrib/yaml/postgresql-10.yaml [hunk 2]
```diff
@@ -0,0 +1,244 @@
+# primary_information()
```
Comment (jesperpedersen) [inline contrib/yaml/postgresql-10.yaml:1]: Start with a comment that this a sample file for PostgreSQL 10

File: src/include/pgexporter.h [hunk 3]
```diff
@@ -117,11 +135,25 @@ struct user
...
+   char name[MISC_LENGTH];          /*< Column name */
+   char description[MISC_LENGTH];   /*< Description of column */
+} __attribute__ ((aligned (64)));
+
 /** @struct
  * Defines the Prometheus metrics
  */
 struct prometheus
 {
+   char query_str[MAX_QUERY_LENGTH];               /*< The query string of metric */
```
Comment (jesperpedersen) [inline src/include/pgexporter.h:153]: Maybe just `query`

File: src/include/pgexporter.h [hunk 4]
```diff
@@ -167,11 +199,14 @@ struct configuration
...
    int number_of_admins;         /**< The number of admins */
+   int number_of_metrics;        /**< The number of metrics*/
+
+   char metrics_path[MISC_LENGTH]; /**< The metrics path */
 
    struct server servers[NUMBER_OF_SERVERS];       /**< The servers */
    struct user users[NUMBER_OF_USERS];             /**< The users */
    struct user admins[NUMBER_OF_ADMINS];           /**< The admins */
-   struct prometheus prometheus;                   /**< The Prometheus metrics */
+   struct prometheus prometheus[NUMBER_OF_METRICS];/**< The Prometheus metrics */
```
Comment (jesperpedersen) [inline src/include/pgexporter.h:209]: We can look at this - maybe there is just 1 `yaml_queries` in `prometheus` instead... Keep as is for now

File: src/include/yamlConfiguration.h [hunk 5]
```diff
@@ -0,0 +1,30 @@
+/*
```
Comment (jesperpedersen) [inline src/include/yamlConfiguration.h:1]: Rename file to `yaml_configuration.h`.

Add file header.

File: src/libpgexporter/yamlConfiguration.c [hunk 6]
```diff
@@ -0,0 +1,273 @@
+
```
Comment (jesperpedersen) [inline src/libpgexporter/yamlConfiguration.c:1]: Rename file to `yaml_configuration.c`.

Add file header.

File: src/libpgexporter/yamlConfiguration.c [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+   int status = START_STATUS;
+   int idx_metrics = 0;
+   int idx_columns = 0;
+   int in_columns = 0;
+
+   file = fopen(filename, "r");
+   if (file == NULL)
+   {
+      printf("fopen() error");
+      return -1;
```
Comment (jesperpedersen) [inline src/libpgexporter/yamlConfiguration.c:31]: Use logging, and `1` as a return value

File: src/libpgexporter/yamlConfiguration.c [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+   }
+   if (!yaml_parser_initialize(&parser))
+   {
+      printf("yaml_parser_initialize() error");
+      return -1;
```
Comment (jesperpedersen) [inline src/libpgexporter/yamlConfiguration.c:36]: Ditto

**Comment by jesperpedersen** [comment]
I think this pull request is very close - although I havn't verified the actual functionality yet :)

Make sure that all queries are valid or logging has been performed in order to notify the user that something is wrong, and where.

**Comment by xRay2016** [comment]
I feel sorry for forgetting to reply :sob:. 
I will fix the code according to the suggestion and improve the `error log`.

**Comment by jesperpedersen** [comment]
Sounds good !

**Comment by jesperpedersen** [comment]
This is looking good, but we need some updates for

```
.github/workflows/ci.yml
doc/CONFIGUTATION.md
doc/YAML.md
```

such that the code compiles in CI, and configuration and format are explained

File: README.md [hunk 9]
```diff
@@ -49,9 +49,10 @@ See [Architecture](./doc/ARCHITECTURE.md) for the architecture of `pgexporter`.
 * [OpenSSL](http://www.openssl.org/)
 * [systemd](https://www.freedesktop.org/wiki/Software/systemd/)
 * [rst2man](https://docutils.sourceforge.io/)
+* [libyaml](https://pyyaml.org/wiki/LibYAML)
 
 ```sh
-dnf install git gcc cmake make libev libev-devel openssl openssl-devel systemd systemd-devel python3-docutils
+dnf install git gcc cmake make libev libev-devel openssl openssl-devel systemd systemd-devel python3-docutils libyaml-devel
```
Comment (jesperpedersen) [inline README.md:55]: Lets add `libyaml` as well

File: doc/CONFIGURATION.md [hunk 10]
```diff
@@ -44,6 +44,7 @@ See a [sample](./etc/pgexporter.conf) configuration for running `pgexporter` on
 | backlog | 16 | Int | No | The backlog for `listen()`. Minimum `16` |
 | hugepage | `try` | String | No | Huge page support (`off`, `try`, `on`) |
 | pidfile | | String | No | Path to the PID file |
+| metrics_path | | String | No | Path to customized metrics |
```
Comment (jesperpedersen) [inline doc/CONFIGURATION.md:47]: Add a comment that it can be both a file or a directory with YAML files

File: doc/YAML.md [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+| tag | | Yes | The tag of metric |
+| sort | name | No | The sort type of metric (name, data) |
+| columns | | Yes | The specific information of columns  | 
+
+
+## columns 
+| Property | Default | Required | Description |
+|----------|---------|----------|-------------|
+| name | | No | The name of this column |
+| type | | Yes | The type of column(label,gauge...) |
```
Comment (jesperpedersen) [inline doc/YAML.md:22]: `counter` is missing, and spacing

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

File: src/libpgexporter/configuration.c [hunk 12]
```diff
@@ -1095,6 +1115,17 @@ pgexporter_reload_configuration(void)
       goto error;
    }
 
+   if (strlen(config->metrics_path) > 0)
```
Comment (jesperpedersen) [inline src/libpgexporter/configuration.c:1118]: This isn't correct - see how config/users/admins are loaded a bit further up.

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by jesperpedersen** [comment]
Pull the `server_type` filter commit off - save it for another pull request, or squash it into the main patch. I would rename it to just `server` but keep the `both`, `primary` and `replica` settings.

I don't see an update to the reload code...

**Comment by xRay2016** [comment]
> Pull the `server_type` filter commit off - save it for another pull request, or squash it into the main patch. I would rename it to just `server` but keep the `both`, `primary` and `replica` settings.
> 
> I don't see an update to the reload code...

OK! I pushed branch wrongly, now it is correct. I update the reload code.

File: src/libpgexporter/configuration.c [hunk 13]
```diff
@@ -1090,6 +1113,17 @@ pgexporter_reload_configuration(void)
       goto error;
    }
 
+   if (strlen(reload->metrics_path) > 0)
```
Comment (jesperpedersen) [inline src/libpgexporter/configuration.c:1116]: Put this below the `admin` reload code, and do it the same way -- there is no `sd_notify` and `exit` here...

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

File: src/libpgexporter/configuration.c (hunk 13 above)
Comment (xRay2016) [reply src/libpgexporter/configuration.c]: OK! I have changed it.

**Comment by jesperpedersen** [comment]
Squashed, and merged - congrats on your first pull request !

Thanks for your contribution !

**Comment by xRay2016** [comment]
Thank you very much for your help!

**Pull request closed without merging**